src/aws-cpp-sdk-core/source/client/AWSClient.cpp

```diff
@@ -1162,3 +1166,44 @@ void AWSClient::AppendRecursionDetectionHeader(std::shared_ptr<Aws::Http::HttpRe
 
     ioRequest->SetHeaderValue(Aws::Http::X_AMZN_TRACE_ID_HEADER, xAmznTraceIdVal);
 }
+
+
+
+AWSError<CoreErrors> AWSClient::BuildAWSErrorFromResponseBody(const std::shared_ptr<Aws::Http::HttpResponse>& httpResponse) const
+{
+
+    Aws::String message{"Error in body of the response"};
+    //extract error message and code in the body
+    auto& body = httpResponse->GetResponseBody();
+    auto responseCode = httpResponse->GetResponseCode();
+    auto readPointer = body.tellg();
+    XmlDocument doc = XmlDocument::CreateFromXmlStream(body);
+    Aws::String bodyError;
+    auto coreErrorCode = CoreErrors::INTERNAL_FAILURE;
+    
+    if (doc.WasParseSuccessful() &&
```

Comment:
this logic should be applied only on S3 Client.

I'd suggest to add this logic into generated [S3ErrorMarshaller](https://github.com/aws/aws-sdk-cpp/blob/main/generated/src/aws-cpp-sdk-s3/include/aws/s3/S3ErrorMarshaller.h#L16)

The base class has a [virtual method `Marshall`](https://github.com/aws/aws-sdk-cpp/blob/a95671dc38b2c61a18f5398d1e549278e1fbe440/src/aws-cpp-sdk-core/source/client/AWSErrorMarshaller.cpp#L94) that could be overwritten.